### PR TITLE
make it possible (on the back end) to run reports off aggregate data sources

### PR DESCRIPTION
**Pull request by czue**
I was quite surprised that this was all it took to get regular UCR `ReportConfig`s wired to aggregate table definitions.

I have mixed feelings about this PR. 

On the one hand it is very small and contained and will unlock a fair amount of value to be able to start exercising aggregate reports against prod data. I also believe it is relatively low-risk in terms of breaking any existing workflows (famous last words).

On the other hand, it relies on the fact that `DataSourceConfiguration` and `AggregateTableDefinition` share a substantial portion of a not-explicitly-declared API, which makes it quite hard to reason with and also not very future-proof. 

I think the right long term thing to do is manipulate the UI side of the code to work with objects that inherit from a base class (which presumably these two classes either inherit from or can be converted into via some kind of [adapter pattern](https://en.wikipedia.org/wiki/Adapter_pattern)). But, going down that route now would require a substantially larger amount of refactors and testing and be much more likely to break existing edge case reports. So, may be overkill until we decide we want to keep this feature around?

Keen to get thoughts from any other interested stakeholders, including @emord @snopoke @nickpell @esoergel et. al.

Note that there is currently no UI to actually configure a report against an aggregate data source, so this would have to be done via shell until I add that.

### CODE REVIEW

**Review by snopoke — COMMENTED** [review]
Pretty cool. I get what you're saying about this being dangerous because there isn't a defined 'interface'. Would it be hard to make both classes extend an abstract base class that defined the interface?

**Comment by czue** [comment]
> Would it be hard to make both classes extend an abstract base class that defined the interface?

I'm actually not sure. I just attempted this and immediately ran into [this issue](https://stackoverflow.com/questions/8723639/a-django-model-that-subclasses-an-abc-gives-a-metaclass-conflict) which looks like it could be a bit involved to sort out but am sure is workable. Have we run into that before and have any way of dealing with it? We could just not use `abc` I suppose.

I have the vague sense that maybe the report UI code shouldn't bother itself with data sources at all but only work with the `IndicatorSqlAdapter` class, which would be a bit cleaner in terms of separation of concerns and also solve the defined interface problem. But I was worried about making that change because I don't think there is great test coverage for all the possible reports.

**Comment by snopoke** [comment]
I think we came across that problem when adding the SQL case classes and
chose to keep the base class but not use ABC:
https://github.com/dimagi/commcare-hq/blob/dea41adc548cb6178ea839a3a8abf48efd08bf86/corehq/form_processor/abstract_models.py#L158

I have the vague sense that maybe the report UI code shouldn't bother
> itself with data sources at all but only work with the IndicatorSqlAdapter
>  class, which would be a bit cleaner in terms of separation of concerns
> and also solve the defined interface problem. But I was worried about
> making that change because I don't think there is great test coverage for
> all the possible reports.


I can't really comment on this since I don't know much about the internals
of how the reports work in terms of how they determine what columns there
are etc. It does seem reasonable though.

**Comment by czue** [comment]
Ok, good call. updated and the process made me realize I could delete some duplicated code that didn't need to be there.

Not sure if there's a better way to handle fields, but I noticed that on `AbstractCommCareCase` we also don't define the shared fields so guessing we don't have one.

**Review by snopoke — APPROVED** [review]

**Review by emord — APPROVED** [review]

File: corehq/apps/userreports/models.py [hunk 1]
```diff
@@ -904,7 +929,7 @@ def bulk_creation(cls, doc_ids, doc_type, domain, config_ids):
         ])
 
 
-def get_datasource_config(config_id, domain):
+def get_datasource_config(config_id, domain, data_source_type=DATA_SOURCE_TYPE_STANDARD):
```
Comment (esoergel) [inline corehq/apps/userreports/models.py]: This is awesome

**Review by esoergel — COMMENTED** [review]
:+1:

**Pull request merged**